### PR TITLE
Basic user-facing Noosphere integration

### PR DESCRIPTION
**Pull request by gordonbrander**
This PR gets Noosphere integration up-to-par for end-user use.

## Changes

- Primary ID used throughout app is now `MemoAddress` instead of `Slug`. `MemoAddress` is a slug plus an `Audience`.
- Update database schema
    - memo primary key is now a string representation of `MemoAddress`. This allows us to index and search across both sphere-managed and local content in the same table.
    - In addition, we have a `slug` and `address` key to allow for querying these properties.
- Introduce NoosphereService, a threadsafe manager of `Noosphere` and `SphereFS` instances.
- New DataService methods for
     - Getting sphere identity and version
     - Syncing sphere to gateway
     - Writing sphere state to database
- Improved first run experience
- Introduce `AppDefaults` which offer type-safe getters/setters to `UserDefaults`.
    - We use this to persist sphere ID and some other book-keeping details.
- Update and clean up First Run Experience
- Get settings pane working

## App behavior updates:

- When launching app:
    - Sync sphere with gateway
        - Then sync sphere to database
    - Sync local files with database
 - When pull-to-refresh
     - Sync sphere with gateway
        - Then sync sphere to database
    - Sync local files with database

## QA Checklist

Tested the following after the refactor:

- [x] Fresh start
- [x] Bad Noosphere gateway
- [x] No Noosphere gateway
- [x] Rebuild DB
- [x] Launch First Run Experience
- [x] Relaunch First Run Experience
- [x] Noosphere sync
- [x] File system sync
- [x] Noosphere pull-to-refresh
    - [x] File system pull-to-refresh
- [x] List notes
- [x] Count notes
- [x] Create note
    - [x] Lists refresh
- [x] Open note
- [x] Update note
    - [x] Lists refresh
- [x] Retitle note
    - [x] Lists refresh
- [x] Move note
    - [x] Lists refresh
- [x] Merge note
    - [x] Lists refresh
- [x] Remove note
    - [x] Lists refresh
- [x] Change note audience
    - [x] Lists refresh
- [x] Search notes
- [x] Click link for note
- [x] Click link for note that does not exist
- [x] Search links
- [x] Add link

### CODE REVIEW

**Comment by gordonbrander** [comment]
Still need to fix this issue. May be on the Subconscious side, rather than the Noosphere side. https://github.com/subconsciousnetwork/noosphere/issues/224

File: xcode/Subconscious/Shared/AppTheme.swift [hunk 1]
```diff
@@ -66,10 +66,9 @@ extension Color {
     static let placeholderText = SwiftUI.Color("PlaceholderTextColor")
     static let textPressed = text.opacity(0.5)
     static let textDisabled = placeholderText
-    static let secondaryText = SwiftUI.Color("SecondaryTextColor")
     static let tertiaryText = placeholderText
     static let icon = SwiftUI.Color.accentColor
-    static let secondaryIcon = SwiftUI.Color.secondaryText
+    static let secondaryIcon = SwiftUI.Color.secondary
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/AppTheme.swift:71]: SwiftUI now exposes secondary color via SwiftUI.Color. We should use that instead.

File: xcode/Subconscious/Shared/Components/AppTabView.swift [hunk 2]
```diff
@@ -19,7 +19,7 @@ struct AppTabView: View {
                 .tabItem {
                     Label("Feed", systemImage: "newspaper")
                 }
-            NotebookView(parent: store)
+            NotebookView(app: store)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppTabView.swift:22]: Converging on a consisten naming scheme where the app store is always called `app`.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 3]
```diff
@@ -85,21 +95,90 @@ final class DatabaseService {
         }
     }
 
+    /// Deletes and rebuilds the database by applying migrations to a new
+    /// database.
+    func rebuild() throws -> Int {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:100]: Rebuild can be used to delete and reconstruct database, recovering from a broken state.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 4]
```diff
@@ -85,21 +95,90 @@ final class DatabaseService {
...
+    /// Write database metadata at type-safe metadata key
+    func writeMetadatadata(key: DatabaseMetaKeys, value: String) throws {
+        try writeMetadatadata(key: key.rawValue, value: value)
+    }
+
     /// Write entry syncronously
-    func writeEntry(
-        entry: MemoEntry,
-        info: FileInfo
+    func writeMemo(
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:160]: Preferring more direct "write memo to address" signatures wherever possible, to match the shape of the Noosphere API.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 5]
```diff
@@ -127,39 +208,41 @@ final class DatabaseService {
...
             parameters: [
-                .text(slug.description)
+                .text(address.description)
             ]
         )
     }
     
     /// Count all entries
-    func countEntries() -> Int? {
+    func countMemos() -> Int? {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:245]: Prefer "memo" terminology wherever it makes sense. We used to use "entry" as a generic term.  Now we try to speak concretely about "memos" (an actual data structure) and "entries" (another data structure). Some APIs still use "entry" in the generic legacy sense. We'll try to fix those as we go.

File: xcode/Subconscious/Shared/Library/SQLite3Database.swift [hunk 6]
```diff
@@ -1,16 +1,6 @@
 import SQLite3
 import Foundation
 
-extension Array {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Library/SQLite3Database.swift:4]: Remove this extension. It assumes too much, and we were relying on it in places we shouldn't have (like tests), causing complection in places we shouldn't be.

File: xcode/Subconscious/Shared/Library/SQLite3Database.swift [hunk 7]
```diff
@@ -163,7 +171,7 @@ final class SQLite3Database {
             text(escapePrefixQueryLike(query))
         }
 
-        public func unwrap() -> String? {
+        public func toString() -> String? {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Library/SQLite3Database.swift:174]: Introduce named `toX` methods. The older `unwrap` methods relied on Swift to disambiguate type information from return type, which it often failed to do, requiring type annotations. This new method naming convention allows for more fluent chaining without type annotations.

File: xcode/Subconscious/Shared/Library/SQLite3Database.swift [hunk 8]
```diff
@@ -209,10 +226,24 @@ final class SQLite3Database {
         }
     }
 
-    public enum SQLite3DatabaseError: Error {
+    public enum SQLite3DatabaseError: Error, LocalizedError {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Library/SQLite3Database.swift:229]: Conform to LocalizedError. This gives us more insight into what goes wrong when there is an error on the SQLite side.

File: xcode/Subconscious/SubconsciousTests/Tests_SphereFS.swift [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+                noosphere: noosphere,
+                identity: sphereIdentity
+            )
+            let version = try sphere.version()
+            
+            XCTAssertEqual(version, endVersion)
+        }
+    }
+    
+    func testWritesThenCloseThenReopenWithVars() throws {
```
Comment (gordonbrander) [inline xcode/Subconscious/SubconsciousTests/Tests_SphereFS.swift:435]: This test fails.

Note I overwrite the old noosphere variable with a new one in the same scope. It is therefore possible that the Noosphere deinit does not get called until the scope exits. deinit is where we deallocate the Noosphere pointer.

File: xcode/Subconscious/SubconsciousTests/Tests_SphereFS.swift [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+        _ = try? sphere.sync()
+        
+        XCTAssertEqual(
+            foo.body.toString(),
+            "Test",
+            "Read current version"
+        )
+    }
+    
+    func testWritesThenCloseThenReopenWithScopes() throws {
```
Comment (gordonbrander) [inline xcode/Subconscious/SubconsciousTests/Tests_SphereFS.swift:379]: This test succeeds.

Note I init each Noosphere class in a separate `do` block. `do` blocks create a scope, guaranteeing that Noosphere's `deinit` will be called when scope exits. `deinit` is where we deallocate the Noosphere pointer.

File: xcode/Subconscious/Shared/Services/Noosphere/Noosphere.swift [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+                throw NoosphereError.foreignError("Unknown")
+            }
+            defer {
+                ns_string_free(errorMessagePointer)
+            }
+            let errorMessage = String.init(cString: errorMessagePointer)
+            throw NoosphereError.foreignError(errorMessage)
+        }
+        return value
+    }
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/Noosphere/Noosphere.swift:167]: Bravo 👏

**Review by bfollington — APPROVED** [review]
Phew, that took a while to read, but probably not as long as it took to write. Great work.

File: xcode/Subconscious/Shared/Components/FirstRun/FirstRunView.swift [hunk 12]
```diff
@@ -144,40 +19,40 @@ struct FirstRunView: View {
                     .frame(width: 128, height: 128)
                 Spacer()
                 VStack(alignment: .leading, spacing: AppTheme.unit3) {
-                    Text("Welcome to Subconscious, a place to garden thoughts and share with others.")
+                    Text("Subconscious is a place to garden thoughts and share with others.")
                     
-                    Text("Subconscious is powered by a decentralized note graph. Your data is yours, forever.")
+                    Text("It's powered by a decentralized note graph, so your data belongs to you.")
```
[the remark this answers is not in the document]
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Components/FirstRun/FirstRunView.swift:24]: Will workshop this in a future PR 👍

File: xcode/Subconscious/Shared/Components/FirstRun/FirstRunView.swift [hunk 13]
```diff
@@ -144,40 +19,40 @@ struct FirstRunView: View {
                     .frame(width: 128, height: 128)
                 Spacer()
                 VStack(alignment: .leading, spacing: AppTheme.unit3) {
-                    Text("Welcome to Subconscious, a place to garden thoughts and share with others.")
+                    Text("Subconscious is a place to garden thoughts and share with others.")
```
[the remark this answers is not in the document]
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Components/FirstRun/FirstRunView.swift:22]: Will workshop this in a future PR 👍

File: xcode/Subconscious/Shared/Components/BacklinksView.swift [hunk 14]
```diff
@@ -61,7 +57,20 @@ struct BacklinksView_Previews: PreviewProvider {
...
+                        title: "The Lee Shore",
+                        excerpt: "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi.",
+                        modified: Date.now
+                    ),
+                    EntryStub(
+                        address: MemoAddress(
+                            formatting: "Loomings",
+                            audience: .public
+                        )!,
+                        title: "Floop",
```
[the remark this answers is not in the document]
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Components/BacklinksView.swift:73]: 😌

**Pull request merged**